File: cpp/src/linked_list.cpp
#include "linked_list.hpp"
#include <stdexcept>

namespace linked_list {

LinkedList::LinkedList() : head_(nullptr), tail_(nullptr), size_(0) {}

LinkedList::~LinkedList() {
    Node* current = head_;
    while (current != nullptr) {
        Node* next = current->next;
        delete current;
        current = next;
    }
}

void LinkedList::pushBack(int value) {
    Node* newNode = new Node{value, nullptr};
    if (head_ == nullptr) {
        head_ = newNode;
        tail_ = newNode;
    } else {
        tail_->next = newNode;
        tail_ = newNode;
    }
    size_++;
}

int LinkedList::get(size_t index) const {
    if (index >= size_) {
        throw std::out_of_range("LinkedList: Index out of range");
    }
    Node* current = head_;
    for (size_t i = 0; i < index; i++) {
        current = current->next;
    }
    return current->value;
}

void LinkedList::deleteNode(size_t index) {
    if (index >= size_) {
        throw std::out_of_range("LinkedList: Index out of range");
    }
    // Remove the first node if the index is 0
    if (index == 0) {
        Node* old_head = head_;
        head_ = head_->next;
        if (head_ == nullptr) tail_ = nullptr;
        delete old_head;
        size_--;
    } else {
        Node* current = head_;
        for (size_t i = 0; i < index - 1; i++) current = current->next;
        Node* old_node = current->next;
        current->next = current->next->next;
        if (current->next == nullptr) tail_ = current;
        delete old_node;
        size_--;
    }
}

size_t LinkedList::size() const {
    return size_;
}

void LinkedList::traverse(std::function<void(int)> const& f) {
    Node* current = head_;
    while (current != nullptr) {
        f(current->value);
        current = current->next;
    }
}

}

File: python/src/linked_list.py
from typing import Callable


class LinkedList:
    class _Node:
        # Slots are used to reduce memory usage by avoiding the creation of a dictionary for each instance
        __slots__ = ("value", "next")

        def __init__(self, value):
            self.value = value
            self.next = None

    def __init__(self):
        self.head = None  # head node
        self.tail = None  # tail node (for fast append)
        self._size = 0

    def push_back(self, value: int):
        new_node = self._Node(value)

        if self.tail is None:
            # Empty list, head and tail both point to new node
            self.head = new_node
            self.tail = new_node
        else:
            # List is not empty, add new node after tail and update tail
            self.tail.next = new_node
            self.tail = new_node

        self._size += 1

    def get(self, index: int) -> int:
        if index < 0 or index >= self._size:
            return None
        current = self.head
        # Traverse the list until we find the node at the given index
        for _ in range(index):
            current = current.next
        return current.value if current else None

    def delete(self, index: int) -> bool:
        if index < 0 or index >= self._size:
            return False
        # Remove the first node if the index is 0
        if index == 0:
            self.head = self.head.next
            if self._size == 1:  # list becomes empty, reset tail
                self.tail = None
            self._size -= 1
            return True
        # Traverse the list until we find the node before the one at the given index
        current = self.head
        for _ in range(index - 1):
            current = current.next if current else None
        # Remove the node at the given index
        if current and current.next:
            if current.next == self.tail:  # deleting last node, update tail
                self.tail = current
            current.next = current.next.next
            self._size -= 1
            return True
        return False

    def traverse(self, f: Callable[[int], None]):
        current = self.head
        while current:
            f(current.value)
            current = current.next

    def size(self) -> int:
        return self._size
